### PR TITLE
In-algorithm checkpoint loading check

### PR DESCRIPTION
**Pull request by Haichao-Zhang**
Added check on missing-keys and unexpected-keys for in-algorithm checkpoint loading.

### CODE REVIEW

File: alf/algorithms/algorithm.py [hunk 1]
```diff
@@ -239,7 +239,14 @@ def _preload_checkpoint(self):
             stat_dict = extract_sub_state_dict_from_checkpoint(
                 checkpoint_prefix, checkpoint_path)
 
-            self.load_state_dict(stat_dict, strict=True)
+            status = self.load_state_dict(stat_dict, strict=True)
```
Comment (emailweixu) [inline alf/algorithms/algorithm.py:242]: does this handle the case of mismatching dimension?

File: alf/algorithms/algorithm.py [hunk 2]
```diff
@@ -239,7 +239,14 @@ def _preload_checkpoint(self):
...
+            assert not status.missing_keys and not status.unexpected_keys, (
+                "\033[1;31m Checkpoint mis-matches with the model: \033[1;0m \n"
```
Comment (emailweixu) [inline alf/algorithms/algorithm.py]: mis-matches => mismatches

File: alf/algorithms/algorithm.py (hunk 1 above)
Comment (Haichao-Zhang) [reply alf/algorithms/algorithm.py:242]: Yes. The original code can already handle the case of mismatching dimensions. I also add a test case for it now in the second commit.
What happened for the case met yesterday is like what happened in ``3)`` of ``test_mismatch_checkpoint ``, where all the parameters are skipped.

File: alf/algorithms/algorithm.py (hunk 2 above)
Comment (Haichao-Zhang) [reply alf/algorithms/algorithm.py]: Fixed

File: alf/algorithms/algorithm_test.py [hunk 3]
```diff
@@ -394,6 +394,62 @@ def test_partial_preloading_and_then_checkpoint_loading(self):
...
+            ckpt_mngr_composed = ckpt_utils.Checkpointer(
+                ckpt_dir_composed, alg=alg_composed)
+            ckpt_mngr_composed.save(0)
+
+            ckpt_path = ckpt_dir_composed + '/ckpt-0'
+
+            # 3）Checkpoint loading with a missing key case by specifying
+            # a wrong prefix (``alg._sub_alg3``). Test Exception will be raised
+            # and missing key message will be generated.
+            with self.assertRaises(Exception) as context:
```
Comment (emailweixu) [inline alf/algorithms/algorithm_test.py]: `assertRaises(AssertionError)` otherwise some real error will be ignored.

Comment (Haichao-Zhang) [reply alf/algorithms/algorithm_test.py]: Changed

**Review by emailweixu — APPROVED** [review]

**Pull request merged**